### PR TITLE
Constraint contours with rounded-off edgepath bug fix

### PR DESCRIPTION
**Pull request by etpinard**
Fixes a `contourcarpet` bug reported in https://github.com/plotly/phoenix-integration/issues/244

before: https://codepen.io/etpinard/pen/BXwzvw
after: https://codepen.io/etpinard/pen/bXoeOa

This bug also affects `contour` traces with `type: 'constraint'` contours. See

before: https://codepen.io/etpinard/pen/QeqEZm?editors=0010
after: https://codepen.io/etpinard/pen/jgGrex?editors=1010

This PR is another step towards fixing all the _edge_ cases at the contour _edges_ - like https://github.com/plotly/plotly.js/pull/1309

---

This bug occurs when a `contour(carpet)` trace with `contours.type: 'constraint'` has all its `edgepaths` (computed in [`findAllPaths`](https://github.com/plotly/plotly.js/blob/master/src/traces/contour/find_all_paths.js)) rounded off (done [here in `findAllPaths`](https://github.com/plotly/plotly.js/blob/dbe942adcf6c7c439d661e9c025d48b562229a0e/src/traces/contour/find_all_paths.js#L205)). In this case, the [checks for below the contour value](https://github.com/plotly/plotly.js/blob/dbe942adcf6c7c439d661e9c025d48b562229a0e/src/traces/contour/close_boundaries.js#L39-L64) are off and the resulting contour-fill paths are missing a  [boundary prefix](https://github.com/plotly/plotly.js/blob/dbe942adcf6c7c439d661e9c025d48b562229a0e/src/traces/contour/plot.js#L117-L121) making the fill region incorrect and leading to this bug. 

Thank you very much to whoever wrote these comments 

https://github.com/plotly/plotly.js/blob/dbe942adcf6c7c439d661e9c025d48b562229a0e/src/traces/contour/empty_pathinfo.js#L45-L49

they got me thinking about a potential fix.

The most important part of this PR are the :five: new mocks. Please note that mocks `contour_constraints_equal_boundary_minmax` and `carpet_rounded-off-edgepath-gt` render fine on `master` currently. I added them here as they proved very useful in figuring out how to fix this bug w/o causing regressions.

In brief,

- commit https://github.com/plotly/plotly.js/commit/0fe878f341514c456b4e86e8c2f9ed202eab90a8, rearranges a few things to make `levels` and `constraint` contours logic more similar. This commit should not introduce any changes in behaviour.
- commit https://github.com/plotly/plotly.js/commit/6ba77c695c3c2a1368bbd5d45089813f5410d2f2 is the minimal fix. I'm very confident this commit will need **not** lead to any regressions.
- commits https://github.com/plotly/plotly.js/commit/26e66feafe7f1a7199bb3bb57027c870d180c879 and https://github.com/plotly/plotly.js/commit/6f9a59db21f7bb6e34a8bedd4aaa6980e2c35080 attempt to merge the `closeBoundaries` logic for `contour` and `contourcarpet` together. I'm fairly confident the work here is correct, but perhaps we shouldn't release these commits in a patch version.

Moreover, since I spent almost a full week in `src/traces/contour/` and `src/traces/contourcarpet/`, I made a few clean-up commits and I decided to "might as well" checked off https://github.com/plotly/plotly.js/issues/1311.

cc @plotly/plotly_js

### CODE REVIEW

File: src/traces/contour/find_all_paths.js [hunk 1]
```diff
@@ -83,14 +83,16 @@ function makePath(pi, loc, edgeflag, xtol, ytol) {
         pts.push(getInterpPx(pi, loc, marchStep));
         loc[0] += marchStep[0];
         loc[1] += marchStep[1];
+        locStr = loc.join(',');
 
         // don't include the same point multiple times
         if(equalPts(pts[pts.length - 1], pts[pts.length - 2], xtol, ytol)) pts.pop();
-        locStr = loc.join(',');
```
Comment (archmoj) [inline src/traces/contour/find_all_paths.js]: This is not related to this PR. But I was wondering an error may happen in case of trying to access a negative element if/when `pts.length - 2 < 0`?

Comment (etpinard) [reply src/traces/contour/find_all_paths.js]: That should never happen. `pts` has `length=1` before entering this loop and there's a `pts.push` a few lines above this statement.

Comment (archmoj) [reply src/traces/contour/find_all_paths.js]: @etpinard Thanks for checking that.

**Comment by etpinard** [comment]
(finally) merging this thing for 1.50.0
